File: Java/02-动态数组/src/com/zeng/ArrayList.java
package com.zeng;

@SuppressWarnings("unchecked")
public class ArrayList<E> {
	/**
	 * 元素的数量
	 */
	private int size;
	/**
	 * 所有的元素
	 */
	private E[] elements;
	
	private static final int DEFAULT_CAPACITY = 10;
	private static final int ELEMENT_NOT_FOUND = -1;
	
	public ArrayList(int capaticy) {
		
		capaticy = (capaticy < DEFAULT_CAPACITY) ? DEFAULT_CAPACITY : capaticy;
		elements = (E[]) new Object[capaticy];
		
	}
	
	public ArrayList() {
		this(DEFAULT_CAPACITY);
	}
	
	/**
	 * 清除所有元素
	 */
	public void clear() {
		// 清除内存
		for (int i = 0; i < size; i++) {
			elements[i] = null;
		}
		
		size = 0;
		System.gc();
	}

	/**
	 * 元素的数量
	 * @return
	 */
	public int size() {
		return size;
	}

	/**
	 * 是否为空
	 * @return
	 */
	public boolean isEmpty() {
		return size == 0;
	}

	/**
	 * 是否包含某个元素
	 * @param element
	 * @return
	 */
	public boolean contains(E element) {
		return indexOf(element) != ELEMENT_NOT_FOUND;
	}
	
	/**
	 * 在index位置插入一个元素
	 */
	public void add(int index, E element) {
		
		rangeCheckForAdd(index);
		
		ensureCapacity(size + 1);
		
		for (int i = size; i > index; i --) {
			elements[i] = elements[i - 1];
		}
		
		elements[index] = element;
		size++;
	}

	/**
	 * 添加元素到尾部
	 * @param element
	 */
	public void add(E element) {
		add(size, element);
	}

	/**
	 * 获取index位置的元素
	 * @param index
	 * @return
	 */
	public E get(int index) {
		
		rangeCheck(index);
		return elements[index];
	}

	/**
	 * 设置index位置的元素
	 * @param index
	 * @param element
	 * @return 原来的元素ֵ
	 */
	public E set(int index, E element) {
		
		rangeCheck(index);
		
		E old = elements[index];
		elements[index] = element;
		return old;
	}


	/**
	 * 删除index位置的元素
	 * @param index 下标位置
	 * @return 返回被删除的元素
	 */
	public E remove(int index) {
		
		rangeCheck(index);
		
		E old = elements[index];
		for (int i = index + 1; i < size; i++) {
			elements[i - 1] = elements[i];
		}
		
		// 清空最后一个元素
		elements[--size] = null;
		return old;
	}

	/**
	 * 查看元素的索引
	 * @param element
	 * @return
	 */
	public int indexOf(E element) {
		
		if (element == null) {
			
			for (int i = 0; i < size; i++) {
				if (elements[i] == null) return i;
			}
			
		} else {
			
			for (int i = 0; i < size; i++) {
				if (element.equals(elements[i])) return i;
			}
			
		}
		
		return ELEMENT_NOT_FOUND;
	}
	
	@Override
	public String toString() {
		StringBuilder string = new StringBuilder();
		string.append("size = ").append(size).append(", [");
		
		for (int i = 0; i < size; i++) {
			
			if (i != 0) string.append(", ");
			
			string.append(elements[i]);
		}
		
		string.append("]");
		
		return string.toString();
	}

	private void outOfBounds(int index) {
		throw new IndexOutOfBoundsException("Index:" + index + ", Size" + size);
	}
	
	private void rangeCheck(int index) {
		if (index < 0 || index >= size) outOfBounds(index);
	}
	
	private void rangeCheckForAdd(int index) {
		if (index < 0 || index > size) outOfBounds(index);
	}
	
	private void ensureCapacity(int capacity) {

		int oldCapacity = elements.length;
		if (oldCapacity >= capacity) return;
			
		int newCapacity = oldCapacity + (oldCapacity >> 1);
		E[] newElements = (E[]) new Object[newCapacity];
		
		for (int i = 0; i < size; i++) {
			newElements[i] = elements[i];
		}
		
		elements = newElements;
	}
	
}

File: Swift/02-动态数组/02-动态数组/ArrayList.swift
//
//  ArrayList.swift
//  02-动态数组
//
//  Created by Franklin on 2019/10/22.
//  Copyright © 2019 Franklin Tsang. All rights reserved.
//

enum SomeError: Error {
    case illegalArg(String)
    case outOfBounds(Int, Int)
    case outOfMemory
}

struct ArrayList<E: Comparable> {
    // MARK: - 私有属性
    /// 默认容量
    private var DEFAULT_CAPACITY: Int { 10 }
    /// 元素无法找到
    private var ELEMENT_NOT_FOUND: Int { -1 }
    
    // MARK: - 公共接口
    /// 是否为空
    var isEmpty: Bool { size == 0 }
    
    private var elements: [E?] = [E?]()
    /// 元素的数量
    private(set) var size: Int = 0
    
    init() {
        for _ in 0..<DEFAULT_CAPACITY {
            elements.append(nil)
        }
    }
    
    init(capacity: Int) {
        let capacity = capacity < DEFAULT_CAPACITY ? DEFAULT_CAPACITY : capacity
        for _ in 0..<capacity {
            elements.append(nil)
        }
    }
    
    /// 是否包含某个元素
    /// - Parameter element: 元素
    func contains(_ element: E?) -> Bool {
        return indexOf(element) == ELEMENT_NOT_FOUND ? false : true
    }
    
    /// 添加元素到最后面
    /// - Parameter element: 元素
    mutating func add(_ element: E?)  {
        add(size, element)
    }
    
    /// 往index位置添加元素
    /// - Parameter index: 下标位置
    /// - Parameter element: 添加的元素
    mutating func add(_ index: Int, _ element: E?)  {
        if index < 0 || index > size { return }
        
        ensureCapacity(size + 1)
        
        var i = size
        while i > index {
            elements[i] = elements[i - 1]
            i -= 1
        }
        
        elements[index] = element
        size += 1
        
        
    }
    
    /// 返回index位置对应的元素
    /// - Parameter index: 下标位置
    @discardableResult func get(_ index: Int) -> E? {
        if index < 0 || index >= size { return nil }
        return elements[index];
    }
    
    /// 设置index位置的元素
    /// - Parameter index: 下标位置
    @discardableResult mutating func set(_ index: Int, element: E?) -> E? {
        if index < 0 || index >= size { return nil }
        let old = elements[index]
        elements[index] = element
        return old
    }
    
    /// 删除index位置对应的元素
    /// - Parameter index: 下标位置
    /// - Returns: 被删除的元素
    @discardableResult mutating func remove(_ index: Int) -> E? {
        if index < 0 || index >= size { return nil }
        
        let old = elements[index]
        for i in (index + 1)..<size {
            elements[i - 1] = elements[i]
        }
        size -= 1
        elements[size] = nil
        return old
    }
    
    /// 查看元素的位置
    /// - Parameter element: 元素
    /// - Returns: 元素所在位置
    @discardableResult func indexOf(_ element: E?) -> Int {
        if element == nil {
            for index in 0..<size {
                if elements[index] == nil { return index }
            }
        } else {
            for index in 0..<size {
                if element! == elements[index] { return index }
            }
        }
        return ELEMENT_NOT_FOUND
    }
    
    /// 清除所有元素
    mutating func clear() {
        for i in 0..<size {
            elements[i] = nil
        }
        size = 0
    }
    
    // MARK: - 私有接口
    private func outOfBounds(_ index: Int) throws {
        throw SomeError.outOfBounds(index, size)
    }
    
    private func rangeCheckForAdd(_ index: Int) throws {
        if index < 0 || index > size {
            throw SomeError.outOfBounds(index, size)
        }
    }
    
    /// 扩容
    /// - Parameter capacity: 当前容量
    private mutating func ensureCapacity(_ capacity: Int) {
        let oldCapacity = elements.count
        if oldCapacity >= capacity { return }
        
        // 新容量为旧容量的1.5倍
        let newCapacity = oldCapacity + (oldCapacity >> 1)
        var newElements = [E?]()
        for _ in 0..<newCapacity {
            newElements.append(nil)
        }
        
        for i in 0..<size {
            newElements[i] = elements[i]
        }
        
        elements = newElements
        print("\(oldCapacity) 扩容为 \(newCapacity)")
    }
}
